apis/v1/gateway_types.go

```diff
@@ -654,8 +687,53 @@ type FrontendTLSValidation struct {
 	// +kubebuilder:validation:MaxItems=8
 	// +kubebuilder:validation:MinItems=1
 	CACertificateRefs []ObjectReference `json:"caCertificateRefs,omitempty"`
+
+	// FrontendValidationMode defines the mode for validating the client certificate.
+	// There are two possible modes:
+	//
+	// - AllowValidOnly: In this mode, the gateway will accept connections only if
+	//   the client presents a valid certificate. This certificate must successfully
+	//   pass validation against the CA certificates specified in `CACertificateRefs`.
+	// - AllowInvalidOrMissingCert: In this mode, the gateway will accept
+	//   connections even if the client certificate is not presented or fails verification.
```

Comment:
Here in the documentation I think we need to caveat this and enumerate the reasons under which this mode would be used. Can you please help me to better understand the desire to accept invalid certificates, specifically?

Also, do we potentially want to disconnect the two situations from being lumped in with one another? Are there situations where someone would want to express one, but not the other?

Reply:
AllowInvalidOrMissingCert might be used in 2 cases:
* In test environments
* In a production environment, when a client rolls out a new TLS configuration, the gateway should not reject the connection. Instead, it should forward the connection to the backend. This approach helps in ruling out traffic errors caused by an unhealthy backend or other misroutes, thereby narrowing down the issue to the TLS configuration.

 Ad 2 
I don’t think that more granularity is necessary here. In most cases, when a client's certificate is invalid or missing, authorization is delegated to the Backend. Clients who want to introduce stricter validation can achieve that by adding mTLS headers to a route.

Reply:
Gotcha, so would it be fair to say that enabling the permissive mode here is anticipated to be largely, if not only something that would be done in testing environments, or when used in production environments it would only ever be used transiently?

Reply:
So I think Shane is suggesting something like this:
```suggestion
	// - AllowInvalidOrMissingCert: In this mode, the gateway will accept
	//   connections even if the client certificate is not presented or fails verification.
	//
	// This option has significant security risks and should only be used in testing
	// environments or on a temporary basis in non-testing environments.
```

Does that sound right @shaneutt ?

Reply:
Yes, I think at minimum documenting the caveat here is an improvement.

I was pondering about any API mechanic that we could leverage to make this safer, but I think it gets really complex and I don't think there's a lot of great prior art for such a thing. At least changing the language is good, and actually I would go as far as updating the value itself:

```suggestion
	// - AllowInsecureFallback: In this mode, the gateway will accept
	//   connections even if the client certificate is not presented or fails verification.
	//
	// This option has significant security risks and should only be used in testing
	// environments or on a temporary basis in non-testing environments.
```

In this way, even the name suggests the danger.